# PRCP-1003 – Customer Transaction Prediction

## Problem Statement
* Task 1:-Prepare a complete data analysis report on the given data.
* Task 2:-Create a predictive model which will help the bank to identify which customer will make transactions in future.

#### Import Required Libraries

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

#### Load the Dataset

In [2]:
data = pd.read_csv("train.csv")   # update path if needed
data.head()

,ID_code,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
0,train_0,0,8.9255,-6.7863,11.9081,5.0930,11.4607,-9.2834,5.1187,18.6266,...,4.4354,3.9642,3.1364,1.6910,18.5227,-2.3978,7.8784,8.5635,12.7803,-1.0914
1,train_1,0,11.5006,-4.1473,13.8588,5.3890,12.3622,7.0433,5.6208,16.5338,...,7.6421,7.7214,2.5837,10.9516,15.4305,2.0339,8.1267,8.7889,18.3560,1.9518
2,train_2,0,8.6093,-2.7457,12.0805,7.8928,10.5825,-9.0837,6.9427,14.6155,...,2.9057,9.7905,1.6704,1.6858,21.6042,3.1417,-6.5213,8.2675,14.7222,0.3965
3,train_3,0,11.0604,-2.1518,8.9522,7.1957,12.5846,-1.8361,5.8428,14.9250,...,4.4666,4.7433,0.7178,1.4214,23.0347,-1.2706,-2.9275,10.2922,17.9697,-8.9996
4,train_4,0,9.8369,-1.4834,12.8746,6.6375,12.2772,2.4486,5.9405,19.2514,...,-1.4905,9.5214,-0.1508,9.1942,13.2876,-1.5121,3.9267,9.5031,17.9974,-8.8104


#### Basic Checks

In [3]:
# Shape of Dataset
data.shape

(200000, 202)

In [4]:
# print all columns
data.columns

Index(['ID_code', 'target', 'var_0', 'var_1', 'var_2', 'var_3', 'var_4',
       'var_5', 'var_6', 'var_7',
       ...
       'var_190', 'var_191', 'var_192', 'var_193', 'var_194', 'var_195',
       'var_196', 'var_197', 'var_198', 'var_199'],
      dtype='object', length=202)

In [5]:
# data types
data.dtypes

ID_code     object
target       int64
var_0      float64
var_1      float64
var_2      float64
            ...   
var_195    float64
var_196    float64
var_197    float64
var_198    float64
var_199    float64
Length: 202, dtype: object

In [6]:
# tail
data.tail()

,ID_code,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
199995,train_199995,0,11.4880,-0.4956,8.2622,3.5142,10.3404,11.6081,5.6709,15.1516,...,6.1415,13.2305,3.9901,0.9388,18.0249,-1.7939,2.1661,8.5326,16.6660,-17.8661
199996,train_199996,0,4.9149,-2.4484,16.7052,6.6345,8.3096,-10.5628,5.8802,21.5940,...,4.9611,4.6549,0.6998,1.8341,22.2717,1.7337,-2.1651,6.7419,15.9054,0.3388
199997,train_199997,0,11.2232,-5.0518,10.5127,5.6456,9.3410,-5.4086,4.5555,21.5571,...,4.0651,5.4414,3.1032,4.8793,23.5311,-1.5736,1.2832,8.7155,13.8329,4.1995
199998,train_199998,0,9.7148,-8.6098,13.6104,5.7930,12.5173,0.5339,6.0479,17.0152,...,2.6840,8.6587,2.7337,11.1178,20.4158,-0.0786,6.7980,10.0342,15.5289,-13.9001
199999,train_199999,0,10.8762,-5.7105,12.1183,8.0328,11.5577,0.3488,5.2839,15.2058,...,8.9842,1.6893,0.1276,0.3766,15.2101,-2.4907,-2.2342,8.1857,12.1284,0.1385


In [7]:
# Statistical Summary
data.describe()

,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,var_8,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
count,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,...,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000
mean,0.100490,10.679914,-1.627622,10.715192,6.796529,11.078333,-5.065317,5.408949,16.545850,0.284162,...,3.234440,7.438408,1.927839,3.331774,17.993784,-0.142088,2.303335,8.908158,15.870720,-3.326537
std,0.300653,3.040051,4.050044,2.640894,2.043319,1.623150,7.863267,0.866607,3.418076,3.332634,...,4.559922,3.023272,1.478423,3.992030,3.135162,1.429372,5.454369,0.921625,3.010945,10.438015
min,0.000000,0.408400,-15.043400,2.117100,-0.040200,5.074800,-32.562600,2.347300,5.349700,-10.505500,...,-14.093300,-2.691700,-3.814500,-11.783400,8.694400,-5.261000,-14.209600,5.960600,6.299300,-38.852800
25%,0.000000,8.453850,-4.740025,8.722475,5.254075,9.883175,-11.200350,4.767700,13.943800,-2.317800,...,-0.058825,5.157400,0.889775,0.584600,15.629800,-1.170700,-1.946925,8.252800,13.829700,-11.208475
50%,0.000000,10.524750,-1.608050,10.580000,6.825000,11.108250,-4.833150,5.385100,16.456800,0.393700,...,3.203600,7.347750,1.901300,3.396350,17.957950,-0.172700,2.408900,8.888200,15.934050,-2.819550
75%,0.000000,12.758200,1.358625,12.516700,8.324100,12.261125,0.924800,6.003000,19.102900,2.937900,...,6.406200,9.512525,2.949500,6.205800,20.396525,0.829600,6.556725,9.593300,18.064725,4.836800
max,1.000000,20.315000,10.376800,19.353000,13.188300,16.671400,17.251600,8.447700,27.691800,10.151300,...,18.440900,16.716500,8.402400,18.281800,27.928800,4.272900,18.321500,12.000400,26.079100,28.500700


In [8]:
# Dataset Information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Columns: 202 entries, ID_code to var_199
dtypes: float64(200), int64(1), object(1)
memory usage: 308.2+ MB


In [9]:
# Missing Values Check
data.isnull().sum().head()

ID_code    0
target     0
var_0      0
var_1      0
var_2      0
dtype: int64

In [10]:
# check for duplicate
data.duplicated().sum()

0

In [11]:
# Check data type
data.dtypes

ID_code     object
target       int64
var_0      float64
var_1      float64
var_2      float64
            ...   
var_195    float64
var_196    float64
var_197    float64
var_198    float64
var_199    float64
Length: 202, dtype: object

#### Predictive Model Building

In [12]:
# Feature & Target Separation
X = data.drop(['ID_code', 'target'], axis=1)
y = data['target']

In [13]:
# Print the Feature
X

,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,var_8,var_9,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
0,8.9255,-6.7863,11.9081,5.0930,11.4607,-9.2834,5.1187,18.6266,-4.9200,5.7470,...,4.4354,3.9642,3.1364,1.6910,18.5227,-2.3978,7.8784,8.5635,12.7803,-1.0914
1,11.5006,-4.1473,13.8588,5.3890,12.3622,7.0433,5.6208,16.5338,3.1468,8.0851,...,7.6421,7.7214,2.5837,10.9516,15.4305,2.0339,8.1267,8.7889,18.3560,1.9518
2,8.6093,-2.7457,12.0805,7.8928,10.5825,-9.0837,6.9427,14.6155,-4.9193,5.9525,...,2.9057,9.7905,1.6704,1.6858,21.6042,3.1417,-6.5213,8.2675,14.7222,0.3965
3,11.0604,-2.1518,8.9522,7.1957,12.5846,-1.8361,5.8428,14.9250,-5.8609,8.2450,...,4.4666,4.7433,0.7178,1.4214,23.0347,-1.2706,-2.9275,10.2922,17.9697,-8.9996
4,9.8369,-1.4834,12.8746,6.6375,12.2772,2.4486,5.9405,19.2514,6.2654,7.6784,...,-1.4905,9.5214,-0.1508,9.1942,13.2876,-1.5121,3.9267,9.5031,17.9974,-8.8104
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
199995,11.4880,-0.4956,8.2622,3.5142,10.3404,11.6081,5.6709,15.1516,-0.6209,5.6669,...,6.1415,13.2305,3.9901,0.9388,18.0249,-1.7939,2.1661,8.5326,16.6660,-17.8661
199996,4.9149,-2.4484,16.7052,6.6345,8.3096,-10.5628,5.8802,21.5940,-3.6797,6.0019,...,4.9611,4.6549,0.6998,1.8341,22.2717,1.7337,-2.1651,6.7419,15.9054,0.3388
199997,11.2232,-5.0518,10.5127,5.6456,9.3410,-5.4086,4.5555,21.5571,0.1202,6.1629,...,4.0651,5.4414,3.1032,4.8793,23.5311,-1.5736,1.2832,8.7155,13.8329,4.1995
199998,9.7148,-8.6098,13.6104,5.7930,12.5173,0.5339,6.0479,17.0152,-2.1926,8.7542,...,2.6840,8.6587,2.7337,11.1178,20.4158,-0.0786,6.7980,10.0342,15.5289,-13.9001


In [14]:
# Print the Target
y

0         0
1         0
2         0
3         0
4         0
         ..
199995    0
199996    0
199997    0
199998    0
199999    0
Name: target, Length: 200000, dtype: int64

In [15]:
# Train-Test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, stratify=y
)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((150000, 200), (50000, 200), (150000,), (50000,))

In [16]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
scaler

StandardScaler()

## Model Training

### Logistic Regression

In [17]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(max_iter=1000)
lr.fit(X_train_scaled, y_train)

y_pred_lr = lr.predict(X_test_scaled)
lr

LogisticRegression(max_iter=1000)

### Decision Tree

In [18]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)
dt

DecisionTreeClassifier(random_state=42)

### Random Forest

In [19]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
rf

RandomForestClassifier(random_state=42)

### Gradient Boosting

In [20]:
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(random_state=42)
gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)
gb

GradientBoostingClassifier(random_state=42)

## Model Evaluation

### Evaluation Function

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

def evaluate(y_test, y_pred, model):
    return {
        "Model": model,
        "Accuracy": accuracy_score(y_test, y_pred),
        "Precision": precision_score(y_test, y_pred),
        "Recall": recall_score(y_test, y_pred),
        "F1 Score": f1_score(y_test, y_pred),
        "ROC-AUC": roc_auc_score(y_test, y_pred)
    }

### Model Comparison Table

In [22]:
results = []

results.append(evaluate(y_test, y_pred_lr, "Logistic Regression"))
results.append(evaluate(y_test, y_pred_dt, "Decision Tree"))
results.append(evaluate(y_test, y_pred_rf, "Random Forest"))
results.append(evaluate(y_test, y_pred_gb, "Gradient Boosting"))

results_df = pd.DataFrame(results)
results_df

,Model,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Logistic Regression,0.91328,0.673562,0.265725,0.381102,0.625670
1,Decision Tree,0.83786,0.202126,0.208201,0.205118,0.558198
2,Random Forest,0.89954,1.000000,0.000199,0.000398,0.500100
3,Gradient Boosting,0.90216,0.870787,0.030852,0.059592,0.515170


# Model Comparison Report
----
## Models Evaluated
The following machine learning models were trained and evaluated on the customer transaction dataset:
* Logistic Regression
* Decision Tree
* Random Forest
* Gradient Boosting
All models were evaluated using the same train–test split and consistent evaluation metrics.

---

### Performance Comparison
| Model               | Accuracy    | Precision   | Recall      | F1 Score    | ROC-AUC     |
| ------------------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| Logistic Regression | **0.91328** | **0.67356** | **0.26573** | **0.38110** | **0.62567** |
| Decision Tree       | 0.83786     | 0.20213     | 0.20820     | 0.20512     | 0.55820     |
| Random Forest       | 0.89954     | 1.00000     | 0.00020     | 0.00040     | 0.50010     |
| Gradient Boosting   | 0.90216     | 0.87079     | 0.03085     | 0.05959     | 0.51517     |

---

## Model-wise Performance Analysis
### Logistic Regression
* Achieved the highest ROC-AUC and F1-Score
* Maintained a good balance between precision and recall
* Successfully identified a reasonable number of customers who are likely to make transactions

---

### Decision Tree
* Moderate performance
* Lower accuracy and ROC-AUC compared to Logistic Regression
* More sensitive to data variations and prone to overfitting

---

### Random Forest
* High accuracy but almost zero recall
* Predicted nearly all customers as non-transacting
* ROC-AUC close to 0.5 indicates near-random performance

---

### Gradient Boosting
* High accuracy but very low recall
* Poor detection of customers who will make transactions
* ROC-AUC slightly better than random guessing

---

## Best Model for Production
### Selected Model: Logistic Regression
#### Reason for Selection:
* Highest ROC-AUC (0.62567) and F1-Score (0.38110)
* Best trade-off between identifying transaction and non-transaction customers
* Performs better on imbalanced data compared to other models
* Computationally efficient and easy to deploy in production systems

---

# Report on Challenges Faced and Techniques Used
---
## Challenge 1: Anonymized Feature Names
**Problem**
* The dataset consists of 200 features whose meanings are not provided. Due to anonymization, it was not possible to understand feature importance from a domain perspective or perform traditional exploratory data analysis.

**Technique Used**
* Exploratory Data Analysis (EDA) based on feature interpretation was skipped as per project instruction. Instead, model-based analysis was used, focusing on evaluation metrics such as Accuracy, Precision, Recall, F1-Score, and ROC-AUC.
* Feature-based exploratory analysis was skipped as per project instruction. Instead, multiple machine learning models were trained and evaluated using performance metrics.

**Reason**
* When feature interpretation is not possible, model performance metrics provide a reliable way to assess the predictive capability of the models.

---

## Challenge 2: High Dimensional Data (200 Features)
**Problem**
* The dataset contains a high number of features (200), which increases model complexity and the risk of overfitting, especially for tree-based models.

**Technique Used**
* Multiple algorithms were tested, including Logistic Regression, Decision Tree, Random Forest, and Gradient Boosting. Model comparison was performed using consistent evaluation metrics.

**Reason**
* Testing multiple models helps identify which algorithm performs best when handling high-dimensional anonymized data.

---

## Challenge 3: Class Imbalance in Target Variable
**Problem**
* The target variable is imbalanced, with significantly fewer customers performing transactions. This caused some models to predict mostly the majority class.
* This is evident in the results where Random Forest and Gradient Boosting showed high accuracy but extremely low recall values.

**Technique Used**
* Stratified train-test split was applied to maintain class distribution. Model evaluation focused on Precision, Recall, F1-Score, and ROC-AUC rather than accuracy alone.

**Reason**
* Accuracy can be misleading in imbalanced datasets. Recall and F1-Score better represent the model’s ability to identify customers who will make transactions.

---

## Challenge 4: Poor Minority Class Detection by Ensemble Models Problem
**Problem**
* Random Forest and Gradient Boosting models achieved high accuracy but showed very low recall, indicating failure to identify customers who will perform transactions.

**Technique Used**
* Model comparison was performed to analyze all evaluation metrics. Logistic Regression was selected as the best model due to better balance between precision and recall.

**Reason**
* A model with balanced performance across metrics is more reliable than a model with high accuracy but poor recall in real-world banking applications.

---

## Challenge 5: Model Selection for Production
**Problem**
* Different models showed varying performance across metrics, making model selection challenging.

**Technique Used**
* A model comparison report was created using Accuracy, Precision, Recall, F1-Score, and ROC-AUC. Logistic Regression achieved the highest ROC-AUC (0.6257) and F1-Score (0.3811)

**Reason**
* ROC-AUC and F1-Score are important metrics for imbalanced classification problems, making Logistic Regression the most suitable model for deployment.

---

# Final Conclusion
Despite challenges such as anonymized features, high dimensionality, class imbalance, and inconsistent model behavior, a reliable predictive model was developed. Based on evaluation results, Logistic Regression was selected as the best model due to its superior ability to identify customers likely to make future transactions.